### DISCUSSION TITLE
Github Workflow for SecretDecks

### DISCUSSION THREAD

**Discussion by DanM3rcurius** [SecretDecks/Documentation#4] [Ideas]
## Goal 
Setup initial workflow within github for collaborative creation of document
### Action Items

- [x] Everyone fork the main repository and play around within your forked/private repository by: <editing> (small things), <committing>, <create pull request> (to merge edits into main repository

- [ ] First item to collaborate on should be the Mindfulness Policy and then the Charter!
- [ ] Visualise current workflow (in preparation for contribution policy) @DanM3rcurius 

### Please provide input of your experiences below as comments in this discussion
@evolayak @bUxEE @tamagit80 @stephen-rowan

**Comment by stephen-rowan** [comment]
## How best to contribute on the SecretDeck repository ? ( suggestions for a draft user guide)

- [x] It is a good idea to fork any SecretDeck repository so that a record of your contribution is available to other members to review and comment on.
- [x]  Once you have made a change on your local fork - you can create a pull request to merge your changes with the main repo. 
- [ ] If automatic merge is enabled (for editors and maintainers) it is helpful to provide a comment that summarizes your changes (example : https://github.com/SecretDecks/Documentation/pull/6#issue-649507647).
- [ ] If members decide a merged change needs further thought the file can always be reverted back to an earlier version (this avoids the need for a reviewer). 
- [ ] Members are encouraged to comment on any changes by either 1) going to the pull request (example : https://github.com/SecretDecks/Documentation/pull/6) and commenting below or 2) on a Discussion thread like this one.

**Comment by stephen-rowan** [comment]
## Forking other repos into SecretDecks

My view is that it is fine to fork an outside repo such as [DAO-Open-Source](https://github.com/SecretDecks/DAO-Open-Source) into SecretDecks.

An outside repo may be forked for a couple of reasons -

a) For ease of access to that repo's content locally (in SecretDecks). So it is apparent what changes SecretDecks are contributing to the outside repo.
b) To freeze the repo (to no longer merge with the upstream repo) and customize it locally.

I would suggest - 

1) SecretDecks use the forked [DAO-Open-Source](https://github.com/SecretDecks/DAO-Open-Source) as a local reference copy and a place to submit generic changes back to DAO-Open-Source main (So it becomes a two-way resource).
2) Where SecretDecks need to freely adapt DAO-Open-Source content - to copy/paste to SecretDecks.

Stephen

**Reply by DanM3rcurius** [reply]
> ## Forking other repos into SecretDecks
> My view is that it is fine to fork an outside repo such as [DAO-Open-Source](https://github.com/SecretDecks/DAO-Open-Source) into SecretDecks.
> 
> An outside repo may be forked for a couple of reasons -
> 
> a) For ease of access to that repo's content locally (in SecretDecks). So it is apparent what changes SecretDecks are contributing to the outside repo.
**Perfect, that's what I was thinking!**

> b) To freeze the repo (to no longer merge with the upstream repo) and customize it locally.
**Not sure this is the aim here with SD, ideally the changes that make sense for both also flow in both directions**
> 
> I would suggest -
> 
> 1. SecretDecks use the forked [DAO-Open-Source](https://github.com/SecretDecks/DAO-Open-Source) as a local reference copy and a place to submit generic changes back to DAO-Open-Source main (So it becomes a two-way resource).

**Very much agree.** However might made a mistake 😅 I forked the DAO-O-S repo to my personal, edited sth and then forked it from my private repo to SD😂. Will delete that forked repo (as changes are within the Contribution Survey doc alone) and fork one directly from [DAO-Open-Source](https://github.com/SecretDecks/DAO-Open-Source) into the SecretDecks repo

> 2. Where SecretDecks need to freely adapt DAO-Open-Source content - to copy/paste to SecretDecks.

**Straight from the forked DAO-O-S repo in SD, correct?**
> 
> Stephen

Thank you so much @stephen-rowan 🙏✨

**Reply by stephen-rowan** [reply]
> Straight from the forked DAO-O-S repo in SD, correct?

Yes that right. If you want to take something from DAO-Open-Source and change it to what SecretDecks needs -  just cut/paste at this stage. We can work out a more sophisticated workflow later (e.g. a forked release).

Just remember to retain the right open-source licensing - we both use Apache 2.0 and Creative Commons ( I use CCL attribution for some things).

All the best

**Comment by wlstylr** [comment]
I like this process.  I forked the main repository and have read the Mindfulness policy.   Towards the end, it states that the Mindfulness Policy is to act as the primer for the Charter.  I wasn't sure if the Charter has already been started, or if this first step is to get folks ready to contribute to the Charter?

**Reply by DanM3rcurius** [reply]
Thanks for the feedback and happy to see that you're getting the hang of it so swiftly 😊
After reading your comment I feel the discussion of the mindfulness policy should be the next item (task here: #10), ideally at the same time a trial run for the github workflow. 

The Charter has not yet been drafted, we'll most likely start with the template from @stephen-rowan and the [DAO Open Source](https://github.com/SecretDecks/DAO-Open-Source) and after we reached consensus on the Mindfulness policy.
What do you think? @wlstylr @evolayak @stephen-rowan

**Reply by stephen-rowan** [reply]
Hi wlstylr, Dan

I tend to see the Mindfulness Policy as an informal preparation for a Charter - as a point of reference or a focus for discussion ? I love the way SecretDecks is using such documents to air it's ideas.

I am also working with another Catalyst project that is looking at the early stages of community design and distributed decision making. 

The consensus there is to start simply with a "getting to know each other" session (a Zoom/Discord meeting etc.) where we discuss our values, motivations informally (as an organization grows these can become regular events).

From there I suggest we move onto two key open-source documents. A "Contributions Policy" ([example](https://github.com/SecretDecks/DAO-Open-Source/blob/main/Contribute.md)) and a "Code of Conduct" ([example](https://github.com/SecretDecks/DAO-Open-Source/blob/main/CODE_OF_CONDUCT.md)) . There lots of examples already available online that we can look at. This will help us start to think about how we will work together.

Then we can take the lessons learnt from drafting a Contributions Policy and a Code of Conduct and build a draft Charter. 

In short -

1) We get to know each other
2) We do something together
3) We draft a Charter from lessons learnt from 1 and 2. 

S

**Reply by DanM3rcurius** [reply]
> Hi wlstylr, Dan
> 
> I tend to see the Mindfulness Policy as an informal preparation for a Charter - as a point of reference or a focus for discussion ? I love the way SecretDecks is using such documents to air it's ideas.

Spot on Stephen! Since it's a shorter doc than the charter it poses a good opportunity to try a trial consensus run and take the experience into the Charter design. 
> 
> I am also working with another Catalyst project that is looking at the early stages of community design and distributed decision making.

is there any congruency? Should we connect with them?

> 
> The consensus there is to start simply with a "getting to know each other" session (a Zoom/Discord meeting etc.) where we discuss our values, motivations informally (as an organization grows these can become regular events).

Very good idea! This is sth that we have been doing in a bit un formalised way; hopefully the mindfulness policy will act as a scaffold for a bit of formalisation.

> 
> From there I suggest we move onto two key open-source documents. A "Contributions Policy" ([example](https://github.com/SecretDecks/DAO-Open-Source/blob/main/Contribute.md)) and a "Code of Conduct" ([example](https://github.com/SecretDecks/DAO-Open-Source/blob/main/CODE_OF_CONDUCT.md)) . There lots of examples already available online that we can look at. This will help us start to think about how we will work together.
> 
> Then we can take the lessons learnt from drafting a Contributions Policy and a Code of Conduct and build a draft Charter.
> 
> In short -
> 
> 1. We get to know each other
> 2. We do something together
> 3. We draft a Charter from lessons learnt from 1 and 2.
> 
> S

Makes sense to modularise the charter! 1. Mindfulness Policy -> 2. Contributions Policy -> 3. Code of Conduct (Conduct Policy?) -> Combine and extend in organisation charter. Love it!

**Reply by stephen-rowan** [reply]
Thanks Dan,

There is some congruency - over the next week I am going to be working on a Contributions Policy and Code of Conduct for distributed decision making. What I will do is crossover this to tailor to SecretDecks values/philosophy and everyone can review (like the Know Thyself doc).

**Comment by DanM3rcurius** [comment]
@stephen-rowan What do you think is best practice or makes most sense for us in terms of permission levels? 
I setup a **Founding Core** team and allocated the _Content-writing_ and _Documentation_ repos. I invited all current members contributing in github and gave read rights. Since everything is versioned and we strive for transparency anyhow, I currently don't see a reason to limit the permissions, albeit it need be for sake of scaling & streaming the workflow
@wlstylr you input is of course much appreciated as well!
🙏

**Reply by wlstylr** [reply]
I think complete transparency granted to SecretDeck github members would be ideal.  Is there a policy regarding pull request review before merging?  I've only done a couple pull requests that were for things like typos that I would usually just self approve to be expedient, but asked you to review before merging.  I can imagine this could become annoying as we scale if you end up being the person who needs to approve every pull request, but at the same time I think it may make sense to set a policy that all pull requests be peer reviewed.   

Trust, transparency, and accountability should be paramount.  Any process that obscures these qualities should be tossed.

**Reply by evolayak** [reply]
@wlstylr, @DanM3rcurius yes, transparency and access to everything is a must. 

Re: pull requests, I agree that when more people are onboarded it could get annoying to have someone have to approve every time, with things like typos, grammar, spelling consistency. But for more substantive edits maybe we can include something in the documentation about this (in a style guide)- as in, any substantive edits (copyediting and above) will be submitted as pull requests. At the beginning while there aren't so many of us we could have all the edits peer reviewed. Just some thoughts. 

On that note, I'm happy to draft a simple editorial style guide that everyone here will review and add to as they see fit.

**Reply by stephen-rowan** [reply]
@DanM3rcurius - on transparency and access - I agree with @evolayak and @wlstylr. 

It probably OK to grant rights short of Admin - as this allows deletion of the repo. If you want to democratize Admin rights I suggest we set-up some kind of voting system (e.g. vote 2 or 3 Admins). If we have a single Admin it is a good idea to invite a successor.

As we are pretty savvy we could all have maintainer level rights. But we may want to train new members up from lower access levels.

In general - 

- It is prudent to request that _everyone_ works from forks of the repos (so there is consistency in tracking and ease in reverting everyone's changes).
- It is OK for everyone to merge their own pull requests - as long as comments are left for major changes (small changes such as correcting spelling are usually self-explanatory).
- It should be generally accepted that other members may revert your pull request if they give a good reason (like adhering to a style guide). 
- It is good practice to "semantically" merge pull requests - that is, submit a pull request every time a change that can be described has occurred (this helps people check an audit trail later on).

Specific cases -

- The community might want certain documents to be go through a review process. This can be announced in the document and a reviewer assigned. 
- Some repos like the Token Metadata Registry will enforce reviews or workflows (using GitHub Actions) as part of their functionality - but we will need to program this together.

It is probably a good idea to summarize this GitHub Workflow and incorporate it into the SecretDecks "contributions policy" (a standard open-source doc). So people have just one place to go to learn how to contribute, use GitHub and refer to standards (such as style etc.).

Further details on access levels here (https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-permission-levels-for-an-organization)

Invite a successor (https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/managing-access-to-your-personal-repositories/maintaining-ownership-continuity-of-your-user-accounts-repositories)
